barman/cloud_providers/aws_s3.py

```diff
@@ -783,14 +805,57 @@ def take_snapshot_backup(self, backup_info, instance_identifier, volumes):
             WaiterConfig=self._get_waiter_config(),
         )
 
+        # Apply lock on snapshots if lock mode is specified
+        if self.lock_mode:
+            self._lock_snapshots(
+                snapshots,
+                self.lock_mode,
+                self.lock_duration,
+                self.lock_cool_off_period,
+                self.lock_expiration_date
+            )
+
         backup_info.snapshots_info = AwsSnapshotsInfo(
             snapshots=snapshots,
             region=self.region,
             # All snapshots will have the same OwnerId so we get it from the last
             # snapshot response.
             account_id=snapshot_resp["OwnerId"],
+
         )
 
+    def _lock_snapshots(self, snapshots, lock_mode, lock_duration, lock_cool_off_period, lock_expiration_date):
```

Comment:
This should be revisited and refactored.

- This should lock just one snapshot. 
- The values should not be checked but passed to the the client method call. The checks will be done higher in the code flow, so they will come to this part as they should be. Also, pay attention that if this fails somehow, the backup will also fail. The snapshot will be created though.
- The response of the request to lock the snapshot has all the information you need to output information. Do not manipulate yourself the values in the response object, because if you use `LockDuration` OR `LockExpiresOn`, each one has a different response. If the field is not present in the response object, the lock is successful but the backup fails.
- Use `output.info()` instead of `logging.info()`. This can be imported `from barman import output`. 

**To summarize, try to use this method to lock one snapshot, pass all the arguments, lock the snapshot and output information dynamically from the response object. In **line 768**, where we have `for volume_identifier, volume_metadata in volumes.items():`, you should lock snapshot as they are created.**

**_Also, you should add a mutually exclusive test in the beginning of the backup process for fields that cannot be passed together, like duration and expiration_date OR governance mode with cool_off period. You did it for the client, but its not being checked when using this with a barman server_**



_______________________________________________________________
I can give you two examples of response object that i have from my tests:

`LockDuration`

`{'SnapshotId': 'snap-00adc6c28f422e98c', 'LockState': 'governance', 'LockDuration': 1, 'LockCreatedOn': datetime.datetime(2024, 10, 22, 17, 0, 49, 253000, tzinfo=tzlocal()), 'LockExpiresOn': datetime.datetime(2024, 10, 23, 17, 0, 49, 253000, tzinfo=tzlocal()), 'LockDurationStartTime': datetime.datetime(2024, 10, 22, 17, 0, 49, 253000, tzinfo=tzlocal()), 'ResponseMetadata': {'RequestId': '3f8c3e22-d1b2-47e9-9cb1-f8476b2b2da0', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': '3f8c3e22-d1b2-47e9-9cb1-f8476b2b2da0', 'cache-control': 'no-cache, no-store', 'strict-transport-security': 'max-age=31536000; includeSubDomains', 'vary': 'accept-encoding', 'content-type': 'text/xml;charset=UTF-8', 'transfer-encoding': 'chunked', 'date': 'Tue, 22 Oct 2024 17:00:49 GMT', 'server': 'AmazonEC2'}, 'RetryAttempts': 0}}`

`LockExpiresOn`

`{'SnapshotId': 'snap-0d83456985415270c', 'LockState': 'compliance-cooloff', 'CoolOffPeriod': 3, 'CoolOffPeriodExpiresOn': datetime.datetime(2024, 10, 22, 19, 55, 3, 461000, tzinfo=tzlocal()), 'LockCreatedOn': datetime.datetime(2024, 10, 22, 16, 55, 3, 461000, tzinfo=tzlocal()), 'LockExpiresOn': datetime.datetime(2024, 10, 24, 15, 53, 35, 457000, tzinfo=tzlocal()), 'LockDurationStartTime': datetime.datetime(2024, 10, 22, 16, 55, 3, 461000, tzinfo=tzlocal()), 'ResponseMetadata': {'RequestId': 'bac96fa3-0116-4a1a-9e2e-6808405f8d79', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': 'bac96fa3-0116-4a1a-9e2e-6808405f8d79', 'cache-control': 'no-cache, no-store', 'strict-transport-security': 'max-age=31536000; includeSubDomains', 'vary': 'accept-encoding', 'content-type': 'text/xml;charset=UTF-8', 'transfer-encoding': 'chunked', 'date': 'Tue, 22 Oct 2024 16:55:02 GMT', 'server': 'AmazonEC2'}, 'RetryAttempts': 0}}`